### PR TITLE
Adding Localization support

### PR DESCRIPTION
**Pull request by raphatex**
Hey, 
So I created a new PR and did everything from scratch so that it's clearer.
I decided to go for the gettext() method, here is how I implemented it for now : 

- originals messages + translations are hold in a struct : 
```
- typedef struct {
    char *msg;    // Original message in english
    char *fr;    // FR Translated message
} Translation;

Translation translations[] = {
    {"Folder is empty.", "Le dossier est vide."},
    {"yes", "Oui"},
    {"No ", "Non"},
    {"OK", "OK"},
    {NULL, NULL}  // End of array
};
```

- We have a very simple linear search lookup function that uses `strcmp()` function

```
const char* rg_gettext(const char *text) {
    if (rg_language == 0)
        return text; // in case language == 0 == english -> return the original string

    for (int i = 0; translations[i].msg != NULL; i++) {
        if (strcmp(translations[i].msg, text) == 0) {
            switch (rg_language)
            {
            case 1:
                return translations[i].fr;  // Return the french string
                break;
            }
        }
    }
    return text; // in case no translation exists
}
```

- "Language" setting have been adding in the Global settings menu below "timezone"

Now I have a question : should we make a Translation variable for each language or add another field to the struct like `char *de; // DE Translated message` for example? The first method uses less memory but the second is a bit prettier

### CODE REVIEW

**Comment by raphatex** [comment]
Also I am thinking about creating some python programs to "dump" all the _(string) in one file so that translation is easier and then be able to convert it back in "c" form

**Comment by ducalex** [comment]
> should we make a Translation variable for each language or add another field to the struct like char *de; // DE Translated message for example? 

My intuition would have been to have one language per struct, then you just have a pointer to the correct language array. But I can see your point. Having all languages in the same struct does have some benefits too. The "key" isn't repeated in every file so it's less prone to mistakes. 

It's also nice to see all translations in one file.
````
{"Folder is empty.", "Le dossier est vide.", "<whatever that is in german>"},
````

Downsides would be that if we have more than 2-3 languages it will become difficult to maintain side-by-side. For longer messages it will be pretty hard to keep track on the line. It can't be split in multiple files. 

Though a change of syntax would fix the first few downsides, like if we do declaration on multiple lines it will remain clear for long messages and even if we have 10 languages, eg:
````
{
    .en = "Folder is empty blah blah blah blah blah blah blah blah .",
    .fr = "Le dossier est vide. blah blah blah blah blah blah blah blah blah blah ", 
    .de = "<whatever that is in german>blah blah blah blah blah blah blah blah>",
},
````




So I think I'm fine either way, one language per struct or all languages in one struct. Do what you think is best, just make sure it's easy to edit/understand for future translators!

File: components/retro-go/rg_system.c [hunk 1]
```diff
@@ -416,6 +416,7 @@ rg_app_t *rg_system_init(int sampleRate, const rg_handlers_t *handlers, const rg
     if (enterRecoveryMode)
     {
         rg_display_init();
+        rg_settings_init();
```
Comment (ducalex) [inline components/retro-go/rg_system.c:419]: I really don't want settings to be loaded during recovery, because settings could be the cause of the previous crash.

But I also understand why a localized menu would be desirable so we'll have to figure something out.

**Comment by ducalex** [comment]
> Also I am thinking about creating some python programs to "dump" all the _(string) in one file so that translation is easier and then be able to convert it back in "c" form

I agree a tool to locate all localized strings (and possibly tell us missing translations) would be great!

File: components/retro-go/rg_system.c (hunk 1 above)
Comment (raphatex) [reply components/retro-go/rg_system.c:419]: yeah I agree, it crash without an sdcard, so not ideal. Maybe we can use a byte in the ESP32 flash to store the current language ?

Comment (ducalex) [reply components/retro-go/rg_system.c]: Indeed I suppose we could store some settings in NVS. But it has some downsides too...

For now I would suggest to comment the added `rg_settings_init();` and put a `// FIXME: Find a way to get the user's language without loading all settings` (or something along those lines).

It's not that big of a deal of the recovery menu is in English (or French, if your target defaults to french) and let's not let this small issue block progress on the PR.

**Comment by ducalex** [comment]
Hey, just a ping to say I have not forgotten/ignored this PR.

I'll be back soon (hopefully) to fully review the the work and continue towards merging it ASAP.

**Comment by raphatex** [comment]
No problem take your time ;)

**Comment by ducalex** [comment]
Thanks for the changes, I don't see any obvious problem remaining :)

I should have time to properly test the patch this weekend, after that it can be merged.

**Comment by ducalex** [comment]
Everything seems to work as intended, nice work!

I've made some changes to the PR, simplifications and catching some edge cases. Can you review and see if there's anything you disagree with (you can revert or we can discuss it)?

Also the lack of diacritics in the fonts is a bit annoying (not your fault). I'll do some tests in adding support for latin-1 I think...

Thanks!

**Comment by raphatex** [comment]
Thanks !
Alright so after some testing everything seems fine to me too !
Your changes on the code are really nice.
Only thing I'm thinking is adding the link to the localization.md in the main Readme but I will let you do it when dev will be merged with main.

Latin-1 would be great ! If you want me to I can try to take a look at it

**Comment by ducalex** [comment]
> Latin-1 would be great ! If you want me to I can try to take a look at it

If you want there is some explanation of how it works for proportional fonts in `fonts/fonts.h`. But it's been so long I forgot the details, I remember it was a bit hit-and-miss to generate a good looking font and I think the tooling was windows only. Might be worth making our own font generator in python...

The Basic font works a bit differently, we just need to add the missing glyphs from the latin file here: https://github.com/dhepper/font8x8 .

But we don't need to wait for the better fonts to merge the PR, I'll do a final read and merge it in a few days :)

**Comment by raphatex** [comment]
Yeah sounds great !

Actually I'm almost done adding Latin-1 support, should I wait for you to merge before starting some commits ?

**Comment by raphatex** [comment]
Here are somes previews :



I had to increase the height of the fonts since the overall height of the font was increased because of capital letters with diacritics ex : É. 
So I was wondering if we needed them ?

Also I tried to add OpenSans16, I really like the numbers and the capitals in this one.
Let me know what you think !

**Comment by ducalex** [comment]
Those all look good indeed!

Since I'll have to test on the various devices I have for any font change and change of size can introduces subtle bugs, I'd prefer it be another PR. You can get started right away on your font PR, I don't think it depends on this one to be merged (though it will merged as soon as I'm at my desk today anyway)


As for ÉÒÙÀ I'll leave it up to you if you want to include them. I agree that we don't have many places where we'd need them.

Maybe I can update the rendering code to fallback to the corresponding bare letter when a diacritics glyph is missing, so that translations can at least use proper spelling even if not shown.
